### PR TITLE
Add tests and better documentation for behavior around `prefix` option

### PR DESCRIPTION
**Pull request by devonestes**
There was some confusing documentation and very little test coverage
around the behavior of the `prefix` option in the `Ecto.Repo` behaviour.
I've added more tests to clarify the current expected behavior, and
documented that behavior more thoroughly so folks aren't surprised by
some of the edge cases that arrise when using that option.

Hopefully this will clear up some of the confusion I had that led me to
post #2918 yesterday.

### CODE REVIEW

File: lib/ecto/repo.ex [hunk 1]
```diff
@@ -414,12 +414,18 @@ defmodule Ecto.Repo do
 
   ## Options
 
+    * `:prefix` - The prefix to run the query on (such as the schema path
+      in Postgres or the database in MySQL). This will only be applied if no
+      `@schema_prefix` has previously been set.
```
Comment (josevalim) [inline lib/ecto/repo.ex]: I think the more precise definition is: "This will be applied to all from and joins in the query that did not have a prefix explicitly given (either via the :prefix option on join/from or via `@schema_prefix` in the schema)".

Comment (josevalim) [reply lib/ecto/repo.ex]: Ping. :)

Comment (devonestes) [reply lib/ecto/repo.ex]: :flushed: updated!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
